R/do_adjustRtime-functions.R

```diff
@@ -582,3 +582,283 @@ adjustRtimeSubset <- function(rtraw, rtadj, subset,
+                             method = c("loess", "gam"),
+                             rt_map,
+                             span = 0.5,
+                             resid_ratio = 3,
+                             zero_weight = 10,
+                             bs = "tp") {
+    model <- .rt_model(method = method,
+                       rt_map, span = span,
+                       resid_ratio = resid_ratio,
+                       zero_weight = zero_weight,
+                       bs = bs)
+    adj <- predict(model, newdata = data.frame(obs = rt_raw))
+    if (is.unsorted(adj, na.rm = TRUE)){
+        warning("Adjusted retention times are not sorted, linear ",
+        "interpolation will be performed for the unsorted data points")
+        adj <- .force_sorted(adj)
+        }
+    idx <- which(rt_raw < min(rt_map$obs))
+    lidx <- length(idx)
+    if (lidx)
+        adj[idx] <- rt_raw[idx] - (rt_raw[lidx + 1L] - adj[lidx + 1L])
+    idx <- which(rt_raw > max(rt_map$obs))
+    if (length(idx))
+        adj[idx] <- rt_raw[idx] - (rt_raw[idx[1L] - 1L] - adj[idx[1L] - 1L])
+    adj
+}
+
+#' @description
+#'
+#' Get a model representing the differences between observed and reference
+#' retention times (parameter `rt_map`). After an initial fit, the model is
+#' re-fitted excluding potential outliers.
+#'
+#' @param rt_map `data.frame` with the observed (column `"obs"`) and reference
+#'     (column `"ref"`) retention time pairs.
+#'
+#' @importFrom stats loess predict resid
+#'
+#' @author Carl Brunius, Philippine Louail
+#'
+#' @noRd
+.rt_model <- function(method = c("loess", "gam"),
+                      rt_map, span = 0.5,
+                      resid_ratio = 3,
+                      zero_weight = 10,
+                      bs = "tp", warnings = FALSE){
+    rt_map <- rt_map[order(rt_map$obs), ]
+    # add first row of c(0,0) to set a fix timepoint.
+    rt_map <- rbind(c(0,0), rt_map)
+    weights <- rep(1, nrow(rt_map))
+    weights[1L] <- zero_weight
+
+    if (method == "gam") {
+        .check_gam_library()
+        model <- mgcv::gam(ref ~ s(obs, bs = bs), weights = weights,
+                           data = rt_map)
+    } else {
+        cw <- tryCatch(
+        model <- loess(ref ~ obs, data = rt_map, span = span,
+                       weights = weights) , warning=function(w) w)
+        if(!inherits(cw, "loess")){
+            warning("Warnings arose when fitting the loess model, this file ",
+                    "might have too few datapoints we advise to relax the ",
+                    "matching parameters, to see the warnings run ",
+                    "'warnings = TRUE'")
+            if (warnings == TRUE)
+                message("The Loess Warnings are: ", cw)
+            }
+    }
+    ## compute outliers
+    SSq <- resid(model)^2
+    meanSSq <- mean(SSq)
+    not_outlier <- (SSq / meanSSq) < resid_ratio
+
+    ## re-run only if there is outliers and keep the zero.
+    if (sum(!not_outlier)){
```

Comment:
maybe use here `if (any(!not_outlier)) {` instead of `sum` - `any` will loop through the logical vector and break the first time it finds a `TRUE` while `sum` will iterate until the end - it's just a very minor performance difference here, but it's good to keep such things in mind too.